libs/ml/src/core/node.cpp

```diff
@@ -0,0 +1,147 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2020 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "ml/core/node.hpp"
+
+namespace fetch {
+namespace ml {
+
+using Shape = fetch::math::SizeVector;
+
+/**
+ * returns the stored operation type and syncs it with operation type of
+ * underlying Ops.
+ * @return
+ */
+template <typename TensorType>
+OpType Node<TensorType>::OperationType()
+{
+  if (operation_type_ != op_ptr_->OperationType())
+  {
+    FETCH_LOG_ERROR(this->name_.c_str(), "Node operation type (" +
+                                             std::to_string(static_cast<int>(operation_type_)) +
+                                             ") and underlying Ops operation code (" +
+                                             std::string(op_ptr_->Descriptor()) + ") mismatch!");
+    operation_type_ = op_ptr_->OperationType();
+  }
+  return operation_type_;
+}
+
+/**
+ * @brief SliceOutputShape computes an output shape of the Node, if only 1 data slice (e.g.
+ * with batch size == 1) is provided to the Graph input. If there is no cached output shape,
+ * the method is recursively called until either a cached shape or input Node is encountered.
+ * @return vector of SizeType.
+ */
+template <typename TensorType>
+Shape Node<TensorType>::SliceOutputShape()
+{
+  // Returned cached shape result if available;
+  Shape const candidate     = op_ptr_->SliceOutputShape();
+  bool const  i_am_subgraph = OperationType() == OpType::LAYER_FULLY_CONNECTED;  // DEBUG! REMOVEME
+  if (!candidate.empty() && !i_am_subgraph)
+  {
+    if (input_nodes_.empty() &&
+        (OperationType() == OpType::OP_PLACEHOLDER || OperationType() == OpType::OP_WEIGHTS))
+    {
+      FETCH_LOG_INFO(name_.c_str(), "Shape deduction reached an leaf node : " + this->name_ + " " +
+                                        op_ptr_->OutputShapeAsString());
+    }
+
+    return candidate;
+  }
+
+  if (input_nodes_.empty())
+  {
+    FETCH_LOG_INFO(name_.c_str(), "Shape deduction reached a Graph leaf : " + this->name_);
+    return candidate;
```

Comment:
I suspect static analysis won't let you do this pattern anyway:
if (X)
{
  return ...
}

if (Y)
{
  return ...
}

in any case it would be clearer if there were else statements

Reply:
I'll double check static analysis behavior in this case.